src/serverWallet.ts

```diff
@@ -85,21 +86,29 @@ export class ServerWeb3Wallet extends Wallet {
   }
 
   private async getNonce(): Promise<BigNumber> {
+    defaultLogger.debug("Tx without nonce, obtaining nonce");
     const transactions = await this.walletStorage.getTransactions(
       await this.getAddress()
     );
     const transactionCount = await this.getTransactionCount();
 
+    let nonce = transactionCount;
+
     const gapNonce = this.findGapNonce(transactions, transactionCount);
     if(gapNonce) {
+      defaultLogger.debug("Found gap nonce " + gapNonce);
       return BigNumber.from(gapNonce);
     }
 
     if(transactions.length) {
-      return BigNumber.from(transactions[transactions.length - 1].nonce + 1);
+      const storedNonce = transactions[transactions.length - 1].nonce + 1;
+      //if stored nonce is lower than transaction count, we didn't store all transactions
+      if(storedNonce > nonce) {
```

Comment:
This logic seems to me the same as one in findGapNonce

Reply:
It seems to me that this will never be triggered

Reply:
Gap is between pending tx. This is for current case on staging where we have tx in db with nonce 114 but account has already confirmed tx 114-125

Reply:
https://github.com/NodeFactoryIo/web3-server-wallet/blob/ee7fedf14c6ace15ce10e5f3ba02c066f76132a0/src/serverWallet.ts#L109 
Isn't this basically the same as that

But I maybe know where the bug is.
If findGapNonce is triggered and checks last transaction nonce and inserts the transaction ncount nonce a gap will be created between them and then the second time it runs it will insert invalid nonce because it will try a gap nonce.
We probably need check that gap nonce is not less than transaction count

Reply:
could you apply fix if you know it?

Reply:
Will try